### PR TITLE
feat(projects,github): reorder board cards inside a column

### PR DESCRIPTION
**Pull request by theBGuy**
Adds keyboard and menu reordering of Projects-board cards within their column, writing the project's own global card order — the order GitHub shows everyone — rather than a client-side view sort. `Alt`/`⌥`+`↑`/`↓` step a card one place, `Alt`/`⌥`+`Home`/`End` send it to the column's ends, and the same four moves are reachable from the card menu and the command palette.

### Positioning model and tests

- Adds `src/features/projects/board-positioning.ts` with `planReorder`, which resolves a direction into GitHub's `afterId` (a `move`/`noop`/`held` plan). Landing above a card means anchoring to that card's *global* predecessor, which may belong to another column since the project order interleaves columns; the module is deliberately import-free so Node's type stripping can load it.
- Adds `scripts/board-positioning.test.mjs` (290 lines, `node:test` only, no dev dependency) pinning every direction against an interleaved three-column fixture: adjacent swaps, board-head `null` anchors, cross-column predecessors, column-end no-ops, truncated-board holds, degenerate single-card columns, and out-of-range indices.
- Extracts `lensSorted` in `src/features/projects/board-model.ts` so the sorting branch and the reorder refusal read the view the same way, and adds the shared `SORTED_VIEW_REASON` / `TRUNCATED_ORDER_REASON` strings used by both the menu and the keyboard announcements.

### Backend command

- Adds `gh_set_item_position` in `src-tauri/src/github/project_item_edits.rs`, wrapping `updateProjectV2ItemPosition` and returning a new `BoardOrder` (`itemIds` + `truncated`) parsed from the mutation payload — transactionally fresh, unlike a follow-up read off lagging replicas.
- Registers the command in `src-tauri/src/lib.rs`, and extends the existing hostile-input test to cover `position_input` plus new tests for null-vs-id `afterId` variables and payload-order preservation.

### Frontend wiring

- `src/lib/git/api/projects.ts` gains `ghSetItemPosition` (passing `afterId` explicitly so `null` reaches the backend as `None`), with `BoardOrder` added to `src/lib/git/types/projects.ts`.
- `src/lib/git/queries/projects.ts` adds `useReorderBoardCard` plus the cache machinery: `reorderBoardItem` splices against the *current* cache, `rechunkPages` keeps original page lengths and metadata, `boardPredecessorId` / `boardAnchorId` derive rollback and write anchors (the write skips archived predecessors, which GitHub refuses as anchors), and `applyBoardOrder` re-asserts the server's returned order. A `"reorder"` variant joins `BoardWriteKind`.
- `src/features/projects/ProjectsBoardPanel.tsx` handles the four Alt chords via `eventToBinding` against a local `REORDER_CHORDS` table, computes optimistic landings (`REORDER_LANDING`), tracks `focusItemId` alongside the focus nonce, and adds a sequence-stamped live region so repeated announcements re-fire for screen readers.
- `src/features/projects/BoardCardMenu.tsx` adds a Position section whose four rows always render in fixed order, disabling in place with the reason on the label; `BoardCard.tsx` emits `data-item-id` and `BoardColumn.tsx` uses it to reject a focus claim landing on the wrong card mid-splice (retries raised from 3 to 6 frames).
- `src/lib/hotkeys/registry.ts` registers `move-card-up/down/top/bottom` palette actions (no default binding — the chord is board-local), and `src/lib/hotkeys/binding.ts` gains `Home`/`End` display names.
- `scripts/check-banned-patterns.mjs` allowlists `ProjectsBoardPanel.tsx` for the `eventToBinding` guard check, with the rationale that it matches fixed Alt chords and already pins focus to a board card.

### Documentation

- `README.md`, `site/src/data/capabilities.ts`, `changelog.d/added-projects-board.md` and the board section of `src/features/help/content.ts` all describe the new moves, including the held cases (a sorting saved view, and downward moves past the loaded end).

## Summary by CodeRabbit

- **New Features**
  - Reorder project board cards within their column using Alt/Option + Arrow keys, Home, or End.
  - Access the same actions from the card menu and command palette.
  - Hold keys to continue moving a card, with changes saved at its final position.
  - Reordering is unavailable in sorted views or when board data is incomplete.

- **Documentation**
  - Added keyboard shortcut guidance and explanations of board ordering behavior.

- **Tests**
  - Added coverage for reordering scenarios, boundaries, incomplete data, and filtered views.

### CODE REVIEW

**Comment by theBGuy** [comment]
## Context for reviewers — P8 keyboard card positioning

Projects v2 epic slice P8. Posting while draft so it's on the record before the review lanes fire.

**1. What this does.** Reorders a card within its column from the keyboard (`Alt`+`↑`/`↓`, `Alt`+`Home`/`End`), the card menu (Move up/down/to top/to bottom), and the command palette. It writes GitHub's project-global item POSITION via `updateProjectV2ItemPosition` — the order GitHub shows everyone. The board already rendered cards in POSITION order; this makes that order writable. Cross-column moves stay P5's job; pointer drag stays P10.

**Deliberate calls (one claim each, with evidence):**

**2. Chords are feature-local, not registry bindings, on purpose.** `Alt`+arrow is matched via `eventToBinding(e)` against a fixed table inside the board's own keydown handler, which already gates on DOM containment (focus is on a card). A rebindable global binding can't express "focus is on a board card", so the four palette actions are `defaultBinding: null` (palette-only). This trips `unguarded-binding-dispatcher`; the allowlist entry in `check-banned-patterns.mjs` is justified because the table is hardcoded (not a user binding) and the board holds no editable/typeahead targets.

**3. `afterId: null` means "move to top"** — documented and live-probed. The move-to-top arms send null, not a first-item anchor.

**4. GitHub refuses an ARCHIVED item as a position anchor** — measured live: `VALIDATION`, "The item to be positioned after is archived and cannot be used to update the position of this item". So the `afterId` math walks back past archived cards to the nearest non-archived predecessor (`loadedOrder()` filters archived; `boardAnchorId` skips them). The cache splice and rollback still include archived items (they hold real cache slots) — the two reads are deliberately split (`boardPredecessorId` vs `boardAnchorId`), commented at the split.

**5. `updateProjectV2ItemPosition` returns transient 500s while COMMITTING the write** — measured live with real GitHub request ids: a single spaced write 500'd the response yet the reorder persisted in the server's POSITION order. Under mutation-dense sequences (the live-drive fired dozens/min) the 500 rate rises (secondary-rate-limit / contention shape); spaced single writes were 3/3 clean. So every write (initial + chase follow-ups) retries with bounded waits (300/800ms): a transient 500 resolves on the idempotent re-issue (same `afterId` = no-op reposition → 200), and only a persistent failure throws → truthful rollback + toast (GitHub's own error, request id preserved). Retries are deliberately shallow so they back off rather than storm a rate limit.

**6. Rapid presses coalesce, they don't queue.** At most one write per card per lens is in flight (`reorderingCards`, keyed by repo+project+lens+item); presses during a write splice optimistically and are picked up by the live write's chase, which re-reads the cache each round trip and writes again until cache and server agree (`REORDER_CHASE_LIMIT` 8). The invariant is convergence, not history: intermediate positions may never be written. A press landing between one burst's chase-exit and its settle is guarded — the settle skips its stale-payload re-assert when the card's cache predecessor no longer matches its payload predecessor, so a long hold can't revert its own tail press.

**7. Settle re-asserts payload order without a refetch** (the P7.1 replica-lag pattern): the mutation payload's `items` connection is transactionally fresh (unlike item reads), so `onSettled` patches every cached lens from it and stale-marks the rest — no settle-time refetch that could return a lagging pre-write board. Exhaustion and error arms fall back to `invalidateProjectBoards`.

**Disclosures / accepted edges:**

- **Filtered-lens interleave**: under a saved view's filter, `order` is the lens's loaded list, so a move is exact within the view and approximate in global interleave (hidden items between the global predecessor and the visible one land below the moved card). Same accepted class as P7.1's filtered inserts; named in a unit test + comment.
- **Sorted-lens hold**: a view that sorts holds reordering ("This view orders cards by its sort") because its columns aren't drawn in POSITION order. Gated by the shared `lensSorted(view)` (used by both the columns expression and the gate). NOT live-drivable — the probe project has no sorted view and view-mutation is owner-gated (P10); covered structurally + by review.
- **Optimistic announce can lead a true exhaustion**: "Moved to N of M" is spoken at press time; on a rare exhaustion the settled position can differ. The visual board is the source of truth; re-announcing post-settle wasn't worth the state.
- **Real-OS-key routing** not separately exercised: CDP-synthetic `Alt`-chords (windowsVirtualKeyCode + modifiers) actuated the app's `altKey` handler end-to-end across dozens of landed moves.

**Docs**: README boards bullet, site capability line, a new help guide section ("Ordering a card inside its column") + the keyboard-summary paragraph, and the existing `added-projects-board.md` fragment extended (no new fragment — the epic's board work is one unreleased added- bullet).

**Live-verified (CDP, GitDesktopTesting probe project):** single-press focus rides the moved card; cross-column moves server-verified byte-for-byte in POSITION order (adjacent, cross-column-up with a Todo anchor, null-to-top); menu Position rows with correct "(already first/last)" parentheticals + noop-disable; menu-route and palette-route focus landing on the moved card; boundary silence (top-card `Alt`+`↑` = noop, no toast); coalescing splices; palette surfaces all four actions only with a card selected. Fixture restored to baseline.

---

_Posted by [GitDesktop](https://gitdesktop.app) — automated agent comment, verify before acting on it._

**Comment by theBGuy** [comment]
🤖 **GitDesktop AI security audit** · `opus` · automated

---

No genuinely exploitable security issues in these changes: the new `gh_set_item_position` command sends a constant GraphQL document with all caller values bound as JSON variables over gh's stdin (fixed argv, no shell), and the frontend additions route ids only into React-escaped text/attributes, `Map`-keyed lookups, and a `data-card-index` selector built from a numeric index.

---

_Posted by [GitDesktop](https://gitdesktop.app) — AI output, verify before acting on it._

**Comment by theBGuy** [comment]
🤖 **GitDesktop AI review** · `opus` · automated

---

## Summary

Adds keyboard/menu/palette reordering of a card inside its board column, writing GitHub's project-global item position with an optimistic splice, a coalescing chase, and a payload-based settle. The shape is careful and the plan arithmetic is well factored and well tested — but the settle's board-wide re-assert silently undoes a *second* card's in-flight reorder (and writes the revert back to GitHub), which I'd fix before merge.

### Correctness

- **blocker** — `useReorderBoardCard` `onSettled`, converged arm (`src/lib/git/queries/projects.ts`): the burst-boundary guard is one card wide (`ctx.itemId`'s predecessor) but the patch it guards is board wide — `writeThroughBoards(..., projectItemsFamilyKey(ctx.repo, ctx.projectId), (data) => applyBoardOrder(data, outcome.order))` re-sorts *every* cached item by the payload's rank. Concurrent reorders of different cards are explicitly allowed (`reorderHeldFor` only tests `w.itemId === itemId`, and the fold key includes `itemId`). Sequence: hold Alt+↑ on card A (its `mutationFn` stays in flight for the whole burst), then arrow to B and press Alt+↑ while A is still flying. A converges first; its payload still carries B's *old* position, so `applyBoardOrder` reverts B's optimistic splice. B's chase then reads that clobbered cache (`cached()` → B's old predecessor ≠ `afterId`), writes B back to where it started, converges, and re-asserts — B's move is silently lost on GitHub, with a stable wrong end state (no invalidation corrects it). Reachable on any board whose loaded items all fit the payload's first 100. Fix: extend the existing guard to the board, not just the card — key `reorderingCards` by board (`JSON.stringify([repo, projectId, query])` → `Set<itemId>`; the fold check becomes `map.get(boardKey)?.has(itemId)`, with add/delete in `mutationFn`'s try/finally deleting the board entry when its set empties) and, in the converged arm, `return` before `writeThroughBoards` when that board's set is non-empty — the still-flying write's own settle owns the reconciliation, exactly as the existing "moved since we converged" arm reasons. Knock-on: reword the `reorderingCards` doc block (it currently says the key is per card *per lens*) and the burst-boundary comment, whose "A false skip (some OTHER card shifted our predecessor) is harmless" line is the assumption this finding contradicts.

- **should-fix** — `planReorder` `"bottom"` (`src/features/projects/board-positioning.ts`): the truncation hold only fires at `index === last`. With `truncated: true` and the card anywhere above the loaded end, `bottom` returns `afterId: column[last]` — the last *loaded* card of the column, not the column's bottom. Concrete case: a board showing **Load more**, cursor on the first of ten loaded cards in a column whose tail is in an unfetched page → Alt+End writes the card above that unloaded tail, and after **Load more** it sits mid-column, not at the bottom. `down` is unaffected (its anchor is always a loaded card). Fix: in the `"bottom"` case return `HELD_TRUNCATED` whenever `truncated`, regardless of `index`; add the missing case to `scripts/board-positioning.test.mjs` beside "to bottom on the column's last card is HELD…". Knock-ons: reword `planReorder`'s doc paragraph ("Moving DOWN or to the BOTTOM of a column whose last card is on screen…"), which states the old rule; no copy change needed — the menu's `bottom` row already carries `truncatedReason: "load more first"` and `TRUNCATED_ORDER_REASON` already covers the keyboard route.

- **should-fix** — exhaustion by *failure* is silent (`useReorderBoardCard.mutationFn`, `src/lib/git/queries/projects.ts`): a follow-up `retryWrite()` that fails all three attempts returns `null` → `{ kind: "exhausted" }`, which resolves the mutation successfully. `onError` never runs, no `toastError` fires, and `onSettled` just invalidates — so after a genuine outage mid-burst the card visibly snaps back to the server's position with no explanation, while the same failure on the *initial* write throws and toasts. Fix: carry the cause — `| { kind: "exhausted"; error?: unknown }`, set it from `lastError` on the follow-up-failure return (leave it absent for the round-limit return), and in the exhausted arm `if (outcome.error !== undefined) toastError(outcome.error);` before `invalidateProjectBoards`. Knock-on: the `ReorderOutcome` doc comment claims the two exhaustion routes "settle identically" — that now holds only for the cache half, so reword it.

- **should-fix** — a refused reorder reaches screen readers only (`reorderCard`, `src/features/projects/ProjectsBoardPanel.tsx` — `announce(held)` and `announce(TRUNCATED_ORDER_REASON)`): the sole consumer is the `sr-only` `role="status"` span. A sighted keyboard user who presses Alt+↑ under a sorted view, on a read-only `project` scope, or past the loaded end gets nothing at all — no toast, no row, nothing — and the four palette entries stay listed and enabled (`active && liveCursor !== null`) while held, so invoking **Move board card up** there is a visible no-op. That cuts against the repo's explain-disabled-actions rule, which the menu's own held rows follow. Fix: in the two refusal arms pair the announce with a short terminal `toast(...)` (`toast` is already imported at line 17); keep the `noop` arm silent as it is now, since the card is already at that end.

### Tests

- **should-fix** — the position arithmetic that actually reaches the cache is untested. `planReorder` gets twenty cases, but `reorderBoardItem` (splice indices computed against `flat` and compared against a `rest`-relative `to`), `rechunkPages` (page-length preservation + identity reuse), `applyBoardOrder` (the all-ids-present bail and the stable sort), and `boardItemPredecessor` (last-occurrence lookup + archived skipping, the `positionable` split behind `boardPredecessorId`/`boardAnchorId`) have none — and they're where the archived/duplicate edge cases live. Fix: move those four into an import-free sibling (e.g. `src/lib/git/queries/board-order.ts`, keeping `InfiniteData`/`BoardItem`/`BoardOrder` as `import type`, which Node's stripping erases) and add a `scripts/board-order.test.mjs` in the same shape as `board-positioning.test.mjs`; `src/lib/git/queries/projects.ts` then imports them.

### Documentation

- **should-fix** — two reader-facing claims go stale with this change (site capability line and the changelog fragment are both correct; the site has no boards `FeatureRow` to extend):
  - `README.md` ~line 592 — "Adding, moving, editing a draft, converting, archiving and removing all need write access to the board and the `project` sign-in scope" no longer enumerates reordering; add it.
  - `src/features/help/content.ts`, projects section closing paragraph — "Every write here (adding, moving, editing a draft, converting, archiving, removing) needs the full `project` scope" has the same gap; add reordering there too.
  - `src/features/help/content.ts`, new *Ordering a card inside its column* section — "On a board that's still paging in, the downward moves say **Load more cards to move past the loaded end** instead" sits in the paragraph about greyed-out *rows*, but no row shows that sentence: `REORDER_ROWS` renders "Move down (load more first)" / "Move to bottom (load more first)", and the full sentence only reaches the sr-only live region on the keyboard route. Recast to quote the row's real parenthetical and attribute the long sentence to the keyboard route.

### Nits

- `src/features/help/content.ts`, *Keyboard* paragraph: "Hold {{key:alt}} with those same four keys" follows a sentence naming six keys (↑/↓, ←/→, Home/End), and Alt+←/→ does nothing — name them ("with ↑ / ↓ or Home / End").
- `NO_REORDER` (`ProjectsBoardPanel.tsx`) makes all four rows `noop`, so a menu whose card the board no longer draws reads "Move up (already first)" *and* "Move down (already last)" at once; a single held row with a "this card is no longer on the board" reason would read straight.
- `BoardOrder.truncated` is parsed in `parse_board_order`, declared in `src/lib/git/types/projects.ts`, and never read — `applyBoardOrder` uses the stronger `flat.some((item) => !rank.has(...))` test instead.
- `SORTED_VIEW_REASON`'s doc comment (`board-model.ts`) ends "…for {@link CARD_WRITE_REASON}'s reason", which doesn't parse; say "for the same reason `CARD_WRITE_REASON` lives here".
- `menuTarget` is set on every card pointerdown and never cleared on menu close, so `menuPos`/`reorderPlansFor` (a `loadedOrder()` pass plus four `planReorder` calls) run on every panel render from the first click onward — gating them on `menuBusy` would confine the work to an open menu.

---

_Posted by [GitDesktop](https://gitdesktop.app) — AI output, verify before acting on it._

File: src/lib/git/queries/projects.ts [hunk 1]
```diff
@@ -761,6 +765,131 @@ function restoreBoardItem(
...
+  positionable: boolean,
+): string | null | undefined {
+  if (data === undefined) return undefined;
+  const flat = data.pages.flatMap((page) => page.items);
+  const at = flat.findLastIndex((item) => item.itemId === itemId);
+  if (at === -1) return undefined;
+  for (let i = at - 1; i >= 0; i -= 1) {
+    if (!positionable || !flat[i].isArchived) return flat[i].itemId;
+  }
+  return null;
```
[the remark this answers is not in the document]
Comment (theBGuy) [reply src/lib/git/queries/projects.ts]: Confirmed — with an adjacent duplicate occurrence (a write-through append plus a later `Load more` page holding the same membership), the backward scan from the last occurrence could return the target's own id as the `afterId`, which is an invalid anchor. Fixing by skipping candidates whose `itemId` matches the target while scanning backwards. The cache helpers (`boardItemPredecessor`, `reorderBoardItem`, `rechunkPages`, `applyBoardOrder`) are also being extracted to an import-free module with a duplicate-membership test that pins this case. In the next push.

---

_Posted by [GitDesktop](https://gitdesktop.app) — automated agent comment, verify before acting on it._

File: src/features/projects/BoardCardMenu.tsx [hunk 2]
```diff
@@ -10,6 +10,7 @@ import { OptionValue } from "@/features/conversations/ProjectFieldValues";
 import { clipTitleFromText } from "@/lib/clip-title";
 import type { BoardItem } from "@/lib/git/types";
 import type { BoardColumnModel } from "./board-model";
+import type { ReorderDirection, ReorderPlan } from "./board-positioning";
```
[the remark this answers is not in the document]
Comment (theBGuy) [reply src/features/projects/BoardCardMenu.tsx]: Accepted the consistency concern, with a correction on the mechanism. The shared `TRUNCATED_ORDER_REASON` is a full sentence ("Load more cards to move past the loaded end") tuned for the keyboard route's spoken announcement; dropped into a menu parenthetical beside "(already first)" / "(already last)" it reads awkwardly. So rather than import it here, the terse "load more first" becomes its own named constant co-located with `TRUNCATED_ORDER_REASON`, with a comment noting the deliberate terse-menu vs full-announce split — which removes the duplicated literal and keeps the two surfaces' relationship visible. In the next push.

---

_Posted by [GitDesktop](https://gitdesktop.app) — automated agent comment, verify before acting on it._

File: src/features/projects/ProjectsBoardPanel.tsx [hunk 3]
```diff
@@ -1085,6 +1209,97 @@ export function ProjectsBoardPanel({
...
+    const item = column?.items[from.idx];
+    if (column === undefined || item === undefined || projectId === null)
+      return;
+    const held = reorderHeldFor(item.itemId);
+    if (held !== undefined) {
+      announce(held);
+      return;
+    }
+    const plan = planReorder({
+      order: loadedOrder(),
```
[the remark this answers is not in the document]
Comment (theBGuy) [reply src/features/projects/ProjectsBoardPanel.tsx:1282]: Accepted — the keyboard and palette routes will announce the boundary ("Already first" for up/top, "Already last" for down/bottom) to the live region on a noop, matching the menu's disabled rows. Pairing a visible toast only with the HELD reasons (sorted view, read-only `project` scope, truncated tail, same-card write in flight), and keeping the plain boundary announce-only, so a user arrowing into a column end doesn't get a toast on every keypress. In the next push.

---

_Posted by [GitDesktop](https://gitdesktop.app) — automated agent comment, verify before acting on it._

File: src/features/projects/ProjectsBoardPanel.tsx [hunk 4]
```diff
@@ -1085,6 +1209,97 @@ export function ProjectsBoardPanel({
...
+      query: lensQuery,
+    });
+    // The cursor rides the card to where the optimistic splice puts it; the
+    // column's own focus machinery does the rest off the nonce. The moved card's
+    // id travels with it — that splice lands a frame or two later, so until it
+    // does the new index still resolves to the neighbour being swapped past.
+    const idx = REORDER_LANDING[direction](from.idx, column.items.length);
+    setCursor({ col: from.col, idx });
+    setFocusItemId(item.itemId);
+    setFocusNonce((n) => n + 1);
```
[the remark this answers is not in the document]
Comment (theBGuy) [reply src/features/projects/ProjectsBoardPanel.tsx:1321]: Real, and deferring it with a tracked home rather than rushing the fix into this round. Reachability is narrow: it needs a reorder write to fail *all* its retries (a genuine mid-burst outage, now surfaced by a toast) and then a command-palette reorder before the user re-focuses a card — the arrow route reads the focused card from the DOM (`cardAt`), so it stays identity-correct regardless of a stale cursor. The robust fix is an identity-derived cursor (tracking the moved item via `focusItemId` and re-finding its position after settle/rollback), which is a cursor-model change better landed on its own than folded into this PR. If you'd rather it ride here, that identity-cursor is the execution.

---

_Posted by [GitDesktop](https://gitdesktop.app) — automated agent comment, verify before acting on it._

File: src/lib/git/queries/projects.ts [hunk 5]
```diff
@@ -761,6 +765,131 @@ function restoreBoardItem(
...
+function reorderBoardItem(
+  data: InfiniteData<BoardItems, string | null> | undefined,
+  itemId: string,
+  afterId: string | null,
+): InfiniteData<BoardItems, string | null> | undefined {
+  if (data === undefined) return undefined;
+  const flat = data.pages.flatMap((page) => page.items);
+  const at = flat.findLastIndex((item) => item.itemId === itemId);
+  if (at === -1) return data;
+  const rest = flat.filter((_, i) => i !== at);
```
[the remark this answers is not in the document]
Comment (theBGuy) [reply src/lib/git/queries/projects.ts]: Confirmed. `reorderBoardItem` removes only the last-index occurrence (`filter((_, i) => i !== at)`), so an earlier duplicate (write-through append + a later `Load more`) survives and, if the moved copy lands before it, that stale-position copy becomes the last occurrence — exactly what `oneCardPerItem` renders and `boardItemPredecessor` reads. Fixing by filtering *every* occurrence of the `itemId` before reinserting the single moved copy (which dedupes the cache as a correct side effect), covered by a duplicate-membership test in the extracted helpers' suite. In the next push.

---

_Posted by [GitDesktop](https://gitdesktop.app) — automated agent comment, verify before acting on it._

File: src/lib/git/queries/projects.ts [hunk 6]
```diff
@@ -871,6 +1000,286 @@ export function useMoveBoardCard() {
...
+ * having to claim a delimiter is impossible.
+ */
+const reorderingCards = new Set<string>();
+
+const reorderFoldKey = (args: {
+  repo: string;
+  projectId: string;
+  query: string | null;
+  itemId: string;
+}) => JSON.stringify([args.repo, args.projectId, args.query, args.itemId]);
```
[the remark this answers is not in the document]
Comment (theBGuy) [reply src/lib/git/queries/projects.ts:909]: Splitting this: the board-wide settle race it's adjacent to is fixed this round — a concurrent reorder of a *different* card no longer has its optimistic move reverted, because the converged settle now defers its board-wide `applyBoardOrder` re-assert while any reorder is still in flight on that board (tracked per `(repo, projectId)`), letting the still-flying write's own settle own the reconciliation.

The specific cross-*lens* out-of-order you describe — the fold key includes `query`, so switching saved views mid-reorder can start a second concurrent write for the *same* card, and a slower older request can commit last — is deferred with a tracked home. It needs a saved-view switch mid-burst, and the robust fix is per-card serialization that drops `query` from the key while storing the latest desired anchor separately from the lens cache (so a cross-lens press folds without silent loss). I'd rather land that coalescing change on its own than fold it into this round. The board-race half is in the next push.

---

_Posted by [GitDesktop](https://gitdesktop.app) — automated agent comment, verify before acting on it._

File: src/features/projects/ProjectsBoardPanel.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (theBGuy) [reply src/features/projects/ProjectsBoardPanel.tsx:1321]: Tracked as a project follow-up (identity-derived cursor: re-find `focusItemId` in `columns` after optimistic update, rollback, and server-order reconciliation). It's recorded in the project's own backlog rather than a GitHub issue — the owner curates the issue tracker, so I'll leave that call to them. Thanks for the Learning; keeping this thread open until the follow-up lands is fine.

---

_Posted by [GitDesktop](https://gitdesktop.app) — automated agent comment, verify before acting on it._

File: src/lib/git/queries/projects.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (theBGuy) [reply src/lib/git/queries/projects.ts:909]: Tracked as a project follow-up as well (per-card serialization: drop `query` from the fold key, store the latest desired anchor separately from the lens cache so a cross-lens press folds without silent loss). Recorded in the project's backlog rather than a GitHub issue — the owner curates the tracker. Agreed this thread stays open until that change lands.

---

_Posted by [GitDesktop](https://gitdesktop.app) — automated agent comment, verify before acting on it._

**Comment by theBGuy** [comment]
🤖 **GitDesktop AI review** · `opus` · automated

---

Verified the previous findings against the current diff and reviewed the new hunks as first-class code.

### Resolved since last review

- **Board-wide re-assert clobbering a concurrent card's reorder** — `useReorderBoardCard.onSettled` now defers the converged `writeThroughBoards(...applyBoardOrder...)` while any reorder is in flight on the board (`reorderingBoards`, keyed `[repo, projectId]`, incremented only by writes that actually start and decremented in `mutationFn`'s `finally` before `onSettled` runs). I also checked the same-card tail press the old predecessor guard covered: a press landing before the burst's final `cached()` folds (the read and the `finally` are one synchronous block, so there is no gap) and is picked up by the live chase; one landing after it increments and is seen by the guard. `writeThroughBoards`' deferral (`cancelQueries().then()`, projects.ts:482) is microtask-only, so it can't span a later keypress.
- **`planReorder("bottom")` on a truncated board** — now `if (truncated) return HELD_TRUNCATED` regardless of index, with the doc paragraph rewritten to the new rule and a matching case in `scripts/board-positioning.test.mjs`.
- **Silent exhaustion-by-failure** — `ReorderOutcome` carries `error?`, set from `lastError` on the follow-up-failure return only, and the exhausted arm toasts it; the `ReorderOutcome` doc now splits the cache half from the feedback half.
- **Refusals reaching screen readers only** — `reorderCard` now pairs `announce` with `toast` on both refusal arms and announces the boundary ("Already first"/"Already last") on `noop`. (But see the new finding below.)
- **Untested cache arithmetic** — `rechunkPages` / `boardItemPredecessor` / `reorderBoardItem` / `applyBoardOrder` extracted to the import-free `src/lib/git/queries/board-order.ts` with `scripts/board-order.test.mjs`; `quality.yml`'s `node --test "scripts/*.test.mjs"` picks it up with no install. I traced every case against the implementation — including the two dedup cases and the own-duplicate predecessor skip — and they hold. The `items.length === page.items.length` guard added to `rechunkPages` correctly stops a shortened flatten from vacuously reusing a longer page.
- **Stale doc surfaces** — README's scope sentence and `content.ts`'s "Every write here…" sentence both list reordering; the truncation paragraph now quotes the row's real "(load more first)" and attributes the full sentence to the keyboard route.
- **Nits** — `SORTED_VIEW_REASON`'s doc tail, the "those same four keys" keyboard paragraph, the contradictory four-`noop` menu (now one `CARD_GONE_REASON` row), and the duplicated "load more first" literal (now `TRUNCATED_ROW_REASON`, imported by the menu).

### Correctness

- **should-fix** — `reorderCard`, `src/features/projects/ProjectsBoardPanel.tsx:1271-1275` and `:1294-1298`: the new refusal toasts fire once per key **auto-repeat**. `onBoardKeyDown` has no `e.repeat` gate (nor should it — repeat is what drives the burst), so every repeat reaches `reorderCard`, hits the same `reorderHeldFor`/`held` arm, and calls `toast(...)` with no id, which mints a *new* sonner toast each time. Concrete case, and the most reachable one: on a board that still pages, hold `Alt`+`↓` to push a card down its column — once it reaches the loaded end, every further repeat queues another "Load more cards to move past the loaded end". Same for a `read:project` sign-in or a sorted view, where the very first repeat is already held. `src/main.tsx:68` mounts `<Toaster position="bottom-right" closeButton />` with no `visibleToasts` cap, so a one-second hold leaves dozens of identical toasts draining afterwards, each re-announced by sonner's own live region. Fix: give both refusals one stable toast id so sonner updates in place instead of stacking — add a module constant beside `CARD_GONE_REASON`, e.g. `const REORDER_REFUSAL_TOAST_ID = "board-reorder-refused";`, and pass it at both call sites (`toast(held, { id: REORDER_REFUSAL_TOAST_ID })`, `toast(TRUNCATED_ORDER_REASON, { id: REORDER_REFUSAL_TOAST_ID })`); one shared id is right since only one refusal can be on screen at a time. The `announce` calls can stay as they are — `aria-live="polite"` coalesces to the latest.

### Nits

- `src/lib/git/queries/board-order.ts:135` — `applyBoardOrder`'s closing sentence ("Those lenses reconcile on the stale mark {@link rechunkPages}'s caller leaves behind") is collateral from the extraction: `rechunkPages`' caller is `applyBoardOrder` itself, and nothing there marks anything stale. The stale mark comes from `applyBoardOrder`'s own caller, `writeThroughBoards` in `projects.ts` — say that (a plain-text name, since the symbol isn't importable here).
- `src/features/help/content.ts`, *Ordering a card inside its column*: "On a board that's still paging in, the downward moves are greyed with **(load more first)** on the row" over-claims after this round's `bottom` change — **Move to bottom** is greyed for every card there, but **Move down** is greyed only on the column's last card, so a mid-column card's Move down still works. Split the clause, e.g. "…**Move to bottom** is greyed with **(load more first)**, and so is **Move down** on the column's last card".

Verdict: no blocking issues — remaining items are non-blocking; merge when ready

---

_Posted by [GitDesktop](https://gitdesktop.app) — AI output, verify before acting on it._

File: src/features/projects/ProjectsBoardPanel.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (theBGuy) [reply src/features/projects/ProjectsBoardPanel.tsx:1282]: Landed in `a6bf92b7`: the keyboard/palette noop now announces the boundary ("Already first"/"Already last") to the live region, and the held reasons announce + toast. Resolving.

---

_Posted by [GitDesktop](https://gitdesktop.app) — automated agent comment, verify before acting on it._

File: src/lib/git/queries/projects.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (theBGuy) [reply src/lib/git/queries/projects.ts]: Landed in `a6bf92b7`: `reorderBoardItem` now filters every occurrence of the id and reinserts one copy, so a lingering earlier duplicate can't become the rendered last occurrence. The extraction also surfaced a latent `rechunkPages` bug (an element-wise identity check that passed vacuously over a shortened trailing page, reusing its old longer items) — fixed with a length guard, both pinned in `scripts/board-order.test.mjs`. Resolving.

---

_Posted by [GitDesktop](https://gitdesktop.app) — automated agent comment, verify before acting on it._

**Comment by theBGuy** [comment]
🤖 **GitDesktop AI review** · `opus` · automated

---

Verified the three previous findings against the current diff and reviewed the two fix hunks (the shared refusal-toast id, the `applyBoardOrder` doc rewrite) plus the help-copy re-wrap as first-class code. All three are fixed, and the fixes introduce no correctness collateral; the only thing left is a wrap artifact in the help paragraph the fix touched.

### Resolved since last review

- **Refusal toasts stacking per key auto-repeat** — both arms now pass `{ id: REORDER_REFUSAL_TOAST_ID }` (`ProjectsBoardPanel.tsx:1278`, `:1301`) against the module constant at `:269-273`, so sonner updates one toast in place instead of minting one per repeat. One shared id is right here: `reorderHeldFor` and the truncation `held` arm are mutually exclusive on a single press, and the `announce` pair is untouched.
- **`applyBoardOrder`'s closing doc sentence** — `board-order.ts:135-136` now names `writeThroughBoards` (plain text, correctly, since the symbol isn't importable into this import-free module), and the claim checks out: `writeThroughBoards` in `queries/projects.ts` is what invokes this patch and then calls `markProjectBoardsStale`.
- **Help guide over-claiming the greyed downward rows** — `content.ts:2042-2045` now splits **Move to bottom** (greyed for every card on a truncated board) from **Move down** (greyed only on the column's last card), which matches `planReorder`'s `bottom: if (truncated) return HELD_TRUNCATED` vs `down: index === last ? ...` and the menu's `truncatedReason` wiring in `BoardCardMenu.tsx:59-79`.

The two deferred items on this PR — the identity-derived focus cursor for a palette reorder after a rollback, and per-card (lens-agnostic) write serialization — remain recorded decisions with tracked follow-ups; the code still shows the arrow route deriving `from` from the DOM (`ProjectsBoardPanel.tsx:709`), which is what keeps their reachability narrow.

### Readability

- **nit** — `src/features/help/content.ts:2045-2046`, *Ordering a card inside its column*: the re-wrap left `loaded end**. Under a saved view that` as a ~37-char orphan line mid-paragraph, against the ~90-char wrap every other line in this section uses. Re-flow from `**Load more cards to move past the` through the end of the sentence so no line breaks early.

Nothing else to raise.

Verdict: no blocking issues — remaining items are non-blocking; merge when ready

---

_Posted by [GitDesktop](https://gitdesktop.app) — AI output, verify before acting on it._
